clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -1060,238 +1061,331 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
+
+    if (const auto *IL = dyn_cast<IntegerLiteral>(Idx)) {
+      int64_t Val = IL->getValue().getSExtValue();
+      if (Val < 0)
+        return getDefaultBuiltinObjectSizeResult(Type, ResType);
+
+      // The index is 0, so we don't need to take it into account.
+      if (Val == 0)
+        Idx = nullptr;
+    }
+  }
+
+  // Calculate the flexible array member's object size using these formulae
+  // (note: if the calculation is negative, we return 0.):
   //
-  //       struct s {
-  //         unsigned long flags;
-  //         int count;
-  //         int array[] __attribute__((counted_by(count)));
-  //       }
+  //      struct p;
+  //      struct s {
+  //          /* ... */
+  //          int count;
+  //          struct p *array[] __attribute__((counted_by(count)));
+  //      };
   //
-  //   1) bdos of the flexible array itself:
+  // 1) 'ptr->array':
   //
-  //     __builtin_dynamic_object_size(p->array, 1) ==
-  //         p->count * sizeof(*p->array)
+  //    size_t count = (size_t) ptr->count;
   //
-  //   2) bdos of a pointer into the flexible array:
+  //    size_t flexible_array_member_base_size = sizeof (*ptr->array);
+  //    size_t flexible_array_member_size =
+  //            count * flexible_array_member_base_size;
   //
-  //     __builtin_dynamic_object_size(&p->array[42], 1) ==
-  //         (p->count - 42) * sizeof(*p->array)
+  //    return flexible_array_member_size;
   //
-  //   2) bdos of the whole struct, including the flexible array:
+  // 2) '&ptr->array[idx]':
   //
-  //     __builtin_dynamic_object_size(p, 1) ==
-  //        max(sizeof(struct s),
-  //            offsetof(struct s, array) + p->count * sizeof(*p->array))
+  //    size_t count = (size_t) ptr->count;
+  //    size_t index = (size_t) idx;
   //
-  ASTContext &Ctx = getContext();
-  const Expr *Base = E->IgnoreParenImpCasts();
-  const Expr *Idx = nullptr;
-
-  if (const auto *UO = dyn_cast<UnaryOperator>(Base);
-      UO && UO->getOpcode() == UO_AddrOf) {
-    Expr *SubExpr = UO->getSubExpr()->IgnoreParenImpCasts();
-    if (const auto *ASE = dyn_cast<ArraySubscriptExpr>(SubExpr)) {
-      Base = ASE->getBase()->IgnoreParenImpCasts();
-      Idx = ASE->getIdx()->IgnoreParenImpCasts();
-
-      if (const auto *IL = dyn_cast<IntegerLiteral>(Idx)) {
-        int64_t Val = IL->getValue().getSExtValue();
-        if (Val < 0)
-          return getDefaultBuiltinObjectSizeResult(Type, ResType);
-
-        if (Val == 0)
-          // The index is 0, so we don't need to take it into account.
-          Idx = nullptr;
-      }
-    } else {
-      // Potential pointer to another element in the struct.
-      Base = SubExpr;
-    }
-  }
+  //    size_t flexible_array_member_base_size = sizeof (*ptr->array);
+  //    size_t flexible_array_member_size =
+  //            count * flexible_array_member_base_size;
+  //
+  //    size_t index_size = index * flexible_array_member_base_size;
+  //
+  //    return flexible_array_member_size - index_size;
```

Comment:
Reading the code, looks like if `index < 0 || index_size > flexible_array_member_size`, it returns `0`. Could you specify that in the comment as well?

Reply:
I mentioned it above, but I'll put it as part of the calculations (I wanted to keep them nice and neat :-).